File: Modules/UI/Objective-C/Textfield/ASTextFieldView/ASTextFieldView.m
//
//  ASTextFieldView.m
//  MyChat
//
//  Created by Boris Verbitsky on 01.04.2022.
//

#import "ASTextFieldView.h"

@implementation ASTextFieldView {
  UIEdgeInsets _textContainerInset;
}

-(UIEdgeInsets)textContainerInset {
  return _textContainerInset;
}

- (void)setTextContainerInset:(UIEdgeInsets)textContainerInset
{
  BOOL needsUpdate = !UIEdgeInsetsEqualToEdgeInsets(textContainerInset, _textContainerInset);
  if (needsUpdate) {
    _textContainerInset = textContainerInset;
    [self setNeedsLayout];
  }
}

- (CGRect)textRectForBounds:(CGRect)bounds{
  return CGRectMake(
                    bounds.origin.x + _textContainerInset.left,
                    bounds.origin.y + _textContainerInset.top,
                    bounds.size.width - _textContainerInset.left - _textContainerInset.right,
                    bounds.size.height - _textContainerInset.top - _textContainerInset.bottom);
}

-(CGRect) editingRectForBounds:(CGRect)bounds{
  return [self textRectForBounds:bounds];
}

@end

File: Modules/UI/Nodes/Textfields/ASTextFieldView.swift
//
//  ASTextFieldView.swift
//  UI
//
//  Created by Boris Verbitsky on 20.04.2022.
//

import UIKit

public final class ASTextFieldView: UITextField {

    // MARK: Public properties
    private var _textContainerInset: UIEdgeInsets?

    public var textContainerInset: UIEdgeInsets? {
        get {
            return _textContainerInset
        }

        set {
            if _textContainerInset != newValue {
                _textContainerInset = newValue
                setNeedsLayout()
            }
        }
    }

    // MARK: Public Methods
    public override func textRect(forBounds bounds: CGRect) -> CGRect {
        CGRect(x: bounds.origin.x + (_textContainerInset?.left ?? 0),
               y: bounds.origin.y + (_textContainerInset?.top ?? 0),
               width: bounds.size.width - (_textContainerInset?.left ?? 0) - (_textContainerInset?.right ?? 0),
               height: bounds.size.height - (_textContainerInset?.top ?? 0) - (_textContainerInset?.bottom ?? 0))
    }

    public override func editingRect(forBounds bounds: CGRect) -> CGRect {
        textRect(forBounds: bounds)
    }
}
